qiskit/transpiler/passes/mapping/unroller.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""Pass for unrolling a circuit to a given basis."""
+
+import copy
+import networkx as nx
+
+from qiskit.circuit import QuantumRegister, ClassicalRegister
+from qiskit.transpiler._transpilererror import TranspilerError
+from qiskit.transpiler._basepasses import TransformationPass
+
+
+class Unroller(TransformationPass):
+    """
+    Unroll (expand) non-basis, non-opaque instructions recursively
+    to a desired basis, using decomposition rules defined for each instruction.
+    """
+
+    def __init__(self, basis=None):
+        """
+        Args:
+            basis (list[Instruction]): target basis gates to unroll to
+        """
+        super().__init__()
+        self.basis = basis or []
+
+    def run(self, dag):
+        """Expand all op nodes to the given basis.
+
+        If self.basis is empty, the circuit is unrolled down to
+        fundamental (opaque) gates (U, CX).
+
+        Args:
+            dag(DAGCircuit): input dag
+
+        Returns:
+            DAGCircuit: output unrolled dag
+
+        Raises:
+            TranspilerError: if no decomposition rule is found for an op
+        """
+        # Walk through the DAG and expand each non-basis node
+        topological_sorted_list = list(nx.topological_sort(dag.multi_graph))
+        for node in topological_sorted_list:
+            current_node = dag.multi_graph.node[node]
+            if current_node["type"] == "op" and \
+                    current_node["op"].name not in self.basis and \
+                    not getattr(current_node["op"], 'opaque', False):
+                decomposition_rules = current_node["op"]._decompositions
```

Comment:
if `_decompositions` is going to be part of the API, should it be without the underscore?

Reply:
i think only the unroller pass will use it. but i'm fine with it being without underscore. but it's probably better to do it in #1445 as it generates a lot of noise.